s3file/static/s3file/js/s3file.js

```diff
@@ -1,158 +1,112 @@
-'use strict';
+/**
+ * Parse XML response from AWS S3 and return the file key.
+ *
+ * @param {string} responseText - XML response form AWS S3.
+ * @return {string} - Key from response.
+ */
+export function getKeyFromResponse (responseText) {
+  const xml = new globalThis.DOMParser().parseFromString(responseText, 'text/xml')
+  return decodeURI(xml.querySelector('Key').innerHTML)
+}
 
-(function () {
-  function parseURL (text) {
-    var xml = new window.DOMParser().parseFromString(text, 'text/xml')
-    var tag = xml.getElementsByTagName('Key')[0]
-    return decodeURI(tag.childNodes[0].nodeValue)
+/**
+ * Custom element to upload files to AWS S3.
+ *
+ * @extends HTMLInputElement
+ */
+export class S3FileInput extends globalThis.HTMLInputElement {
+  constructor () {
+    super()
+    this.type = 'file'
+    this.keys = []
+    this.upload = null
   }
 
-  function waitForAllFiles (form) {
-    if (window.uploading !== 0) {
-      setTimeout(function () {
-        waitForAllFiles(form)
-      }, 100)
-    } else {
-      window.HTMLFormElement.prototype.submit.call(form)
-    }
+  connectedCallback () {
+    this.form.addEventListener('formdata', this.fromDataHandler.bind(this))
+    this.form.addEventListener('submit', this.submitHandler.bind(this), { once: true })
+    this.form.addEventListener('upload', this.uploadHandler.bind(this))
+    this.addEventListener('change', this.changeHandler.bind(this))
   }
 
-  function request (method, url, data, fileInput, file, form) {
-    file.loaded = 0
-    return new Promise(function (resolve, reject) {
-      var xhr = new window.XMLHttpRequest()
+  changeHandler () {
+    this.keys = []
+    this.upload = null
+  }
 
-      xhr.onload = function () {
-        if (xhr.status === 201) {
-          resolve(xhr.responseText)
-        } else {
-          reject(xhr.statusText)
-        }
-      }
+  /**
+   * Submit the form after uploading the files to S3.
+   *
+   * @param {SubmitEvent} event - The submit event.
+   * @return {Promise<void>}
+   */
+  async submitHandler (event) {
+    event.preventDefault()
+    this.form.dispatchEvent(new window.CustomEvent('upload'))
+    await Promise.all(this.form.pendingRquests)
+    this.form.requestSubmit(event.submitter)
```

Comment:
This handler may not work as expect?  `requestSubmit` does not actually `submit` data  but fires a `submit` event. and `preventDefault` is called,  no data will post to servers?

Reply:
The submit handler is only called once. So the second time the event is fired, it will succeed. I just added a test case for your issue to prevent future regression.